### PR TITLE
warning

### LINKED ISSUE #76: Make not actually using special actions illegal
Based on [this thread](https://boardgamegeek.com/thread/1909662/article/36261169#36261169)

- [x] (1) Is it legal to take the 1/2 terraform step power action and not build a mine? NO!
- [x] (2) Is it legal to take the 1/2 terraform step power action and build a mine on your home planet type (i.e. not use any free steps)? NO!
- [x] (3) Is it legal to take the 1/2 terraform step power action and build a mine on a Gaia Planet (i.e. not use any free steps)? NO!
- [x] (4) Is it legal to take the Gain a Tech Tile QIC action if there is no tech tile you can take? NO!
- [x] (6) Is it legal to take the Gain knowledge/ore/credits power actions if you cannot gain ANY of the reward? NO!
- [x] (7) Is it legal to take the 1 free step or increased range booster tile actions and not actually place a mine (or a gaiaformer, in the case of the range tile)? NO!
- [x] (8) Is it legal to take a PI special action if it is not possible to carry it out at all? NO! -> covered by #77 

More rules to prevent accidental misuse
- [x] don't allow activating range booster and not using it
- [x] don't allow activating step booster and not using it

Warnings
- [x] expensive terraforming (for 2 or 3)
- [x] step action is partially wasted
- [x] don't allow building without step booster if it would save a terraforming step (still allow it, in an edge case, with 15o, you might want to spend o before taking the booster)
- [x] don't allow building without range booster if it would save a QIC and if the range booster is used up (i.e. missing range is odd) 
- [x] Gaia forming with tokens not in area 1 
- [x] Form federation with tokens not in area 1 

Warnings at end of round -> do this for all actions that could be activated?
- [x] step booster not used
- [x] range booster not used

### PR DESCRIPTION
**Pull request by zeitlinger**
Fixes #76

### CODE REVIEW

File: engine/src/available-command.ts [hunk 1]
```diff
@@ -523,10 +546,15 @@ export function possibleSpaceLostPlanet(engine: Engine, player: Player) {
       continue;
     }
 
+    const warnings: BuildWarning[] = [];
+    if (qicNeeded > 0 && p.hasActiveBooster(Resource.TemporaryRange)) {
+      warnings.push("range-booster-not-used");
+    }
```
Comment (coyotte508) [inline engine/src/available-command.ts:552]: I don't think the temporary range booster can be used with lost planet

File: engine/src/available-command.ts [hunk 2]
```diff
@@ -523,10 +546,15 @@ export function possibleSpaceLostPlanet(engine: Engine, player: Player) {
...
+
     spaces.push({
-      building: Building.Mine,
       coordinates: hex.toString(),
       cost: qicNeeded > 0 ? new Reward(qicNeeded, Resource.Qic).toString() : "~",
+      warnings: warnings,
```
Comment (coyotte508) [inline engine/src/available-command.ts:557]:
```suggestion
      warnings,
```

File: engine/src/available-command.ts [hunk 3]
```diff
@@ -523,10 +546,15 @@ export function possibleSpaceLostPlanet(engine: Engine, player: Player) {
       continue;
     }
 
+    const warnings: BuildWarning[] = [];
+    if (qicNeeded > 0 && p.hasActiveBooster(Resource.TemporaryRange)) {
+      warnings.push("range-booster-not-used");
+    }
+
     spaces.push({
-      building: Building.Mine,
```
Comment (coyotte508) [inline engine/src/available-command.ts:527]: Why remove the building?

File: engine/src/player.ts [hunk 4]
```diff
@@ -242,6 +254,14 @@ export default class Player extends EventEmitter {
       return { possible: false };
     }
 
+    const warnings: BuildWarning[] = [];
+    if (
+      addedCost.some((c) => c.type == Resource.Qic && c.count > 0) &&
+      this.hasActiveBooster(Resource.TemporaryRange)
+    ) {
+      warnings.push("range-booster-not-used");
+    }
```
Comment (coyotte508) [inline engine/src/player.ts:263]: wow this is helpful!

File: engine/src/player.ts [hunk 5]
```diff
@@ -276,6 +307,7 @@ export default class Player extends EventEmitter {
       possible: true,
       cost,
       steps,
+      warnings: warnings,
```
Comment (coyotte508) [inline engine/src/player.ts:310]:
```suggestion
      warnings,
```

File: viewer/src/components/Commands.vue [hunk 6]
```diff
@@ -76,12 +79,15 @@ import Engine, {
...
-import { factionVariantBoard } from "@gaia-project/engine/src/faction-boards";
+import {buildingName} from "../data/building";
+import { ButtonData, ButtonWarning, GameContext } from "../data";
+import {eventDesc} from "../data/event";
+import {factionDesc} from "../data/factions";
+import {FactionCustomization} from "@gaia-project/engine/src/engine";
+import {factionVariantBoard} from "@gaia-project/engine/src/faction-boards";
+import {AvailableBuilding} from "@gaia-project/engine/src/available-command";
+import { buildWarnings } from "../data/warnings";
+import { hexMap, passWarning } from "../logic/commands";
```
Comment (coyotte508) [inline viewer/src/components/Commands.vue:90]: Same comment as always

File: viewer/src/components/Commands.vue [hunk 7]
```diff
@@ -368,7 +374,7 @@ export default class Commands extends Vue {
                 booster,
                 needConfirm: true,
                 tooltip: tiles.boosters[booster].map((spec) => eventDesc(new Event(spec))).join("\n"),
-                warning,
+                warning: warning,
```
Comment (coyotte508) [inline viewer/src/components/Commands.vue:377]: is this your editor too? (repeated twice below as well)

File: viewer/src/components/MoveButton.vue [hunk 8]
```diff
@@ -205,13 +214,13 @@ export default class MoveButton extends Vue {
         this.startingHex = hex;
 
         const map: SpaceMap = this.$store.state.gaiaViewer.data.map;
-        const { range, costs } = this.button;
+        const {range, costs} = this.button;
```
Comment (coyotte508) [inline viewer/src/components/MoveButton.vue:217]: editor shenanigans

File: viewer/src/components/MoveButton.vue [hunk 9]
```diff
@@ -205,13 +214,13 @@ export default class MoveButton extends Vue {
...
 
         const highlighted = new Map();
 
         const withinDistance = map.withinDistance(hex, range);
         for (const target of withinDistance) {
-          highlighted.set(target, { cost: costs?.[map.distance(hex, target)] ?? "~" });
+          highlighted.set(target, {cost: costs?.[map.distance(hex, target)] ?? "~"});
```
Comment (coyotte508) [inline viewer/src/components/MoveButton.vue:223]: editor shenanigans

File: viewer/src/components/SpaceHex.vue [hunk 10]
```diff
@@ -16,13 +16,14 @@
...
+          warn: warning(hex) != null,
         },
       ]"
       @click="hexClick(hex)"
     />
-    <text class="sector-name" v-if="isCenter">{{
-      hex.data.sector[0] === "s" ? parseInt(hex.data.sector.slice(1)) : parseInt(hex.data.sector)
-    }}</text>
+    <text class="sector-name" v-if="isCenter"
+      >{{ hex.data.sector[0] === "s" ? parseInt(hex.data.sector.slice(1)) : parseInt(hex.data.sector) }}
```
Comment (coyotte508) [inline viewer/src/components/SpaceHex.vue:25]: editor shenanigans

File: viewer/src/components/SpaceHex.vue [hunk 11]
```diff
@@ -94,15 +97,21 @@ export default class SpaceHex extends Vue {
     return this.$store.state.gaiaViewer.data.map;
   }
 
+  warning(hex: GaiaHex): string {
+    const warnings = this.highlightedHexes.get(hex)?.warnings;
+    return warnings != null && warnings.length > 0 ? warnings?.map(w => buildWarnings[w].text).join(", ") : null;
```
Comment (coyotte508) [inline viewer/src/components/SpaceHex.vue:102]:
```suggestion
    return warnings?.length > 0 ? warnings?.map(w => buildWarnings[w].text).join(", ") : null;
```

If `warnings` is null,  `warnings.length` is `undefined`, and any comparison with a number will return false.

**Review by coyotte508 — CHANGES_REQUESTED** [review]
Most of the comments are formatting, only the two first comments are "real" comments

File: engine/src/available-command.ts (hunk 3 above)
Comment (zeitlinger) [reply engine/src/available-command.ts:527]: it was not used IIUC

File: viewer/src/components/Commands.vue (hunk 7 above)
Comment (zeitlinger) [reply viewer/src/components/Commands.vue:378]: yes, I added that

Comment (zeitlinger) [reply viewer/src/components/Commands.vue:378]: I mean I did that manually..

File: viewer/src/components/MoveButton.vue (hunk 8 above)
Comment (zeitlinger) [reply viewer/src/components/MoveButton.vue:217]: I with the prettier would take care of it..

File: viewer/src/components/MoveButton.vue (hunk 9 above)
Comment (zeitlinger) [reply viewer/src/components/MoveButton.vue:223]: finally found the settings..

File: engine/src/available-command.ts (hunk 3 above)
Comment (zeitlinger) [reply engine/src/available-command.ts:527]: I'm adding more type info as I change code - so this is how I found out.

File: viewer/src/logic/commands.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+      {
+        cost: coord.cost,
+        warnings: coord.warnings,
+      },
+    ])
+  );
+}
+
+export function buttonWarning(message?: string): ButtonWarning | null {
+  return message == null ? null : { title: "Are you sure?", body: [message] };
```
Comment (coyotte508) [inline viewer/src/logic/commands.ts:18]:
```suggestion
  return message && { title: "Are you sure?", body: [message] };
```

Just posting that once as an example of a JS trick: A && B will return A if !A, or B if !!A.

**Pull request merged**